### PR TITLE
[SPARK-45203][SQL][TEST] Join tests in KeyGroupedPartitioningSuite should use merge join hint

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

It's more robust to use join hints to enforce sort-merge join in the tests, instead of setting configs which may be ineffective after more advanced optimizations in the future.

### Why are the changes needed?

make tests more future-proof

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

N/A

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala [hunk 1]
```diff
@@ -44,25 +45,9 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
...
-  override def afterAll(): Unit = {
-    try {
-      super.afterAll()
-    } finally {
-      conf.setConf(V2_BUCKETING_ENABLED, originalV2BucketingEnabled)
-      conf.setConf(AUTO_BROADCASTJOIN_THRESHOLD, originalAutoBroadcastJoinThreshold)
-    }
-  }
+  override def sparkConf: SparkConf = super.sparkConf
+    .set(V2_BUCKETING_ENABLED, true)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:49]: not quite related, but this is a much simpler way to set per-suite configs.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala [hunk 2]
```diff
@@ -1252,27 +1200,21 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
...
 
           // join keys are not the same as the partition keys, therefore SPJ is not triggered.
-          val df = sql(
-            s"""
-               SELECT id, name, i.price as purchase_price, p.item_id, p.price as sale_price
-               FROM testcat.ns.$items i JOIN testcat.ns.$purchases p
-               ON i.arrive_time = p.time ORDER BY id, purchase_price, p.item_id, sale_price
-               """)
-
+          val df = createJoinTestDF(Seq("arrive_time" -> "time"), extraColumns = Seq("p.item_id"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1203]: I switched the output column order here, so the test result changed.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala [hunk 3]
```diff
@@ -1451,12 +1399,7 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
                 partiallyClustered.toString,
             SQLConf.V2_BUCKETING_ALLOW_JOIN_KEYS_SUBSET_OF_PARTITION_KEYS.key ->
                 allowJoinKeysSubsetOfPartitionKeys.toString) {
-            val df = sql("SELECT id, name, i.price as purchase_price, " +
-                "p.item_id, p.price as sale_price " +
-                s"FROM testcat.ns.$items i JOIN testcat.ns.$purchases p " +
-                "ON i.arrive_time = p.time " +
-                "ORDER BY id, purchase_price, p.item_id, sale_price")
-
+            val df = createJoinTestDF(Seq("arrive_time" -> "time"), extraColumns = Seq("p.item_id"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1402]: ditto

**Comment by cloud-fan** [comment]
cc @viirya @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @cloud-fan .

**Comment by dongjoon-hyun** [comment]
Also, cc @sunchao , too.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:49]: +1. Yes, this looks simpler.

**Review by dongjoon-hyun — APPROVED** [review]
+1, looks better to me. (Pending CIs). Thank you, @cloud-fan .

**Review by sunchao — APPROVED** [review]
LGTM pending CI, thanks @cloud-fan !

**Review by viirya — APPROVED** [review]
Looks good to me. Thanks @cloud-fan

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

**Pull request closed without merging**